relayer/pkg/chainlink/txm/txm.go

```diff
@@ -69,135 +81,167 @@ func (txm *starktxm) Name() string {
-					senderAddress:  tx.senderAddress,
-					accountAddress: tx.accountAddress,
-				}
-				txsByAccount[key] = append(txsByAccount[key], tx.call)
-			}
-			txm.lggr.Infow("creating batch", "totalTxCount", txLen, "batchCount", len(txsByAccount))
-
-			// async process of tx batches
-			var wg sync.WaitGroup
-			wg.Add(len(txsByAccount))
-			for key, txs := range txsByAccount {
-				txm.lggr.Debugw("batch details", "accountAddress", key.accountAddress, "senderAddress", key.senderAddress, "txs", txs)
-				go func(senderAddress caigotypes.Hash, accountAddress caigotypes.Hash, txs []caigotypes.FunctionCall) {
-					// fetch key matching sender address
-					key, err := txm.ks.Get(senderAddress.String())
-					if err != nil {
-						txm.lggr.Errorw("failed to retrieve key", "id", senderAddress, "error", err)
-					} else {
-						// parse key to expected format
-						privKeyBytes := key.Raw()
-						privKey := caigotypes.BigToHex(caigotypes.BytesToBig(privKeyBytes))
-
-						// broadcast batch based on account address
-						hash, err := txm.broadcastBatch(ctx, privKey, accountAddress, txs)
-						if err != nil {
-							txm.lggr.Errorw("transaction failed to broadcast", "error", err, "account", accountAddress, "batchTx", txs)
-						} else {
-							txm.lggr.Infow("transaction broadcast", "txhash", hash)
-						}
-					}
-					wg.Done()
-				}(key.senderAddress, key.accountAddress, txs)
+			// parse key to expected format
+			privKeyBytes := key.Raw()
+			privKey := caigotypes.BigToHex(caigotypes.BytesToBig(privKeyBytes))
+
+			// broadcast tx serially - wait until accepted by mempool before processing next
+			hash, err := txm.broadcast(ctx, privKey, tx.accountAddress, tx.call)
+			if err != nil {
+				txm.lggr.Errorw("transaction failed to broadcast", "error", err, "tx", tx.call)
+			} else {
+				txm.lggr.Infow("transaction broadcast", "txhash", hash)
 			}
-			wg.Wait()
-			tick = time.After(utils.WithJitter(txm.cfg.TxSendFrequency()) - time.Since(start))
-		case <-txm.stop:
-			return
 		}
 	}
 }
 
 const FEE_MARGIN uint64 = 115
 
-func (txm *starktxm) broadcastBatch(ctx context.Context, privKey string, accountAddress caigotypes.Hash, txs []caigotypes.FunctionCall) (txhash string, err error) {
+func (txm *starktxm) broadcast(ctx context.Context, privKey string, accountAddress caigotypes.Hash, tx caigotypes.FunctionCall) (txhash string, err error) {
+	txs := []caigotypes.FunctionCall{tx}
+	client, err := txm.client.Get()
+	if err != nil {
+		return txhash, fmt.Errorf("broadcast: failed to fetch client: %+w", err)
+	}
 	// create new account
-	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), txm.client.Gw, caigo.AccountVersion1)
+	account, err := caigo.NewGatewayAccount(privKey, accountAddress.String(), client.Gw, caigo.AccountVersion1)
+	if err != nil {
+		return txhash, fmt.Errorf("failed to create new account: %+w", err)
+	}
+
+	nonce, err := txm.nonce.NextSequence(accountAddress, client.Gw.ChainId)
```

Comment:
Is fetching the nonce that expensive that we need local nonce management? It just does a read against the local RPC node. It's faster to avoid it, but I'm thinking about cases like someone using the same account (e.g. to transfer funds) while the node is running, it would mess up the txm until a reboot.

Would it make sense to add a recovery mechanism in the nonce manager where it would re-fetch the latest nonce from the account (same logic as Register) if it falls behind? Or is that an edge case we don't see happening since the account will only be used by the txm?

Reply:
e.g.

> // TODO: investigate if nonce management is needed (nonce is requested queried by the sdk for now)

I haven't actually seen a doc or anything that led to the decision that we do need nonce management

Reply:
see comment here on why local nonce management is needed: https://github.com/smartcontractkit/chainlink-starknet/pull/230#discussion_r1196561541

the EVM txm also uses this model for the same reason (summary: always requesting nonce from the chain only works if you only have 1 inflight tx at a time)

also the account used by the CL node should only be used for the CL node - the EVM txm also has that constraint, and would require a re-sync if the nonce changed outside of interactions from the CL node

I think there could be edge cases with the nonce manager, but it's honestly hard to say without seeing this run on a live network. testing locally only goes so far when the RPC behaves 🤷🏼